langchain/callbacks/tracers/wandb.py

```diff
@@ -0,0 +1,322 @@
+
+        To use W&B to monitor all LangChain activity, add this tracer like any other langchain callback
+        ```
+        from wandb.integration.langchain import WandbTracer
+        LLMChain(llm, callbacks=[WandbTracer()])
+        # end of notebook / script:
+        WandbTracer.finish()
+        ```.
+        """
+        super().__init__(**kwargs)
+        self._run_args = run_args
+        self._ensure_run(should_print_url=(wandb.run is None))
+
+    @staticmethod
+    def finish() -> None:
+        """Waits for all asynchronous processes to finish and data to upload.
+
+        Proxy for `wandb.finish()`.
+        """
+        wandb.finish()
+
+    def _log_trace_from_run(self, run: "Run") -> None:
+        """Logs a LangChain Run to W*B as a W&B Trace."""
+        self._ensure_run()
+
+        root_span = safely_convert_lc_run_to_wb_span(run)
+        if root_span is None:
+            return
+
+        model_dict = None
+
+        # TODO: Uncomment this once we have a way to get the model from a run
+        # model = safely_get_span_producing_model(run)
+        # if model is not None:
+        #     model_dict = safely_convert_model_to_dict(model)
+
+        model_trace = trace_tree.WBTraceTree(
+            root_span=root_span,
+            model_dict=model_dict,
+        )
+        wandb.run.log({"langchain_trace": model_trace})
+
+    def _ensure_run(self, should_print_url: bool = False) -> None:
+        """Ensures an active W&B run exists.
+
+        If not, will start a new run with the provided run_args.
+        """
+        if wandb.run is None:
+            # Make a shallow copy of the run args, so we don't modify the original
+            run_args = self._run_args or {}  # type: ignore
+            run_args: dict = {**run_args}  # type: ignore
+
+            # Prefer to run in silent mode since W&B has a lot of output
+            # which can be undesirable when dealing with text-based models.
+            if "settings" not in run_args:  # type: ignore
+                run_args["settings"] = {"silent": True}  # type: ignore
+
+            # Start the run and add the stream table
+            wandb.init(**run_args)
+
+            if should_print_url:
+                print_wandb_init_message(wandb.run.settings.run_url)
+
+        with wb_telemetry.context(wandb.run) as tel:
+            tel.feature.langchain_tracer = True
+
+    # Start of required methods (these methods are required by the BaseCallbackHandler interface)
+    @property
+    def always_verbose(self) -> bool:
+        """Whether to call verbose callbacks even if verbose is False."""
+        return True
+
+    def _generate_id(self) -> Optional[Union[int, str]]:
+        """Generate an id for a run."""
+        return None
+
+    def _persist_run(self, run: "Run") -> None:
+        """Persist a run."""
+        try:
+            self._log_trace_from_run(run)
+        except Exception:
```

Comment:
We actually wrap all callbacks in try catches so not strictly necessary to do this: https://github.com/hwchase17/langchain/blob/master/langchain/callbacks/manager.py#L107